### ISSUE TITLE
async RestClient with Fault Tolerance fails to propagate OpenTracing traces

### ISSUE THREAD

**Issue by Ladicek** [quarkusio/quarkus#8650]
**Describe the bug**
I have a project where I want to use MicroProfile Fault Tolerance on a MicroProfile RestClient interface and propagate OpenTracing traces across the HTTP invocation. The interface looks like this:

```java
@RegisterRestClient(baseUri = "http://localhost:8081")
public interface HelloClient {
    @GET
    @Path("/hello")
    @Produces(MediaType.TEXT_PLAIN)
    @Asynchronous
    @Fallback(fallbackMethod = "fallback")
    CompletionStage<String> get();

    default CompletionStage<String> fallback() {
        return completedFuture("Fallback");
    }
}
```

I want the HTTP invocation to be invoked asynchronously, which is why it returns `CompletionStage` (that's per MP RestClient). When I invoke the `get` method and the server returns an error, I want a fallback to apply; that's why the method has a `@Fallback` annotation, and since it returns `CompletionStage`, it also has the `@Asynchronous` annotation (that's per MP Fault Tolerance; if it didn't have `@Asynchronous`, MP Fault Tolerance would treat the method call as synchronous and fault tolerance behaviors would only work for the method invocation itself and not for all the async actions that happen after the method returns). There's no need to add `@Traced`, because per MP OpenTracing, MP RestClient calls are traced automatically.

Unfortunately, both MP RestClient and MP Fault Tolerance provide asynchronous behavior here. MP Fault Tolerance moves the execution to another thread, and then MP RestClient moves the execution to yet another thread. Because of that, OpenTracing traces are not propagated, and eventually what should be 1 trace ends up as 2.

I was able to trace this down to the RESTEasy implementation of MP RestClient, which collects `AsyncInvocationInterceptor`s (MP RestClient way of doing context propagation) in a `ThreadLocal` variable (see `org.jboss.resteasy.microprofile.client.async.AsyncInvocationInterceptorHandler`). The context is captured properly, but it not activated, because the activation code runs on a different thread and won't find the `Collection<AsyncInvocationInterceptor>`, because it was stored into a `ThreadLocal` on a different thread.

(Note that the problem specifically _is not_ that the HTTP client would set a bad HTTP header with the trace info. The HTTP header is set correctly, and the trace is propagated just fine across the HTTP invocation. The problem is between RestClient capturing the context and then activating it -- between those two moments, Fault Tolerance moves execution to a different thread.)

**Expected behavior**
OpenTracing traces are propagated correctly across async RestClient invocations with Fault Tolerance.

**Actual behavior**
OpenTracing traces are _not_ propagated correctly across async RestClient invocations with Fault Tolerance.

**To Reproduce**
1. `git clone https://github.com/Ladicek/quarkus-restclient-faulttolerance-opentracing-issue.git`
2. `cd quarkus-restclient-faulttolerance-opentracing-issue`
3. `mvn clean test`

Note that the test will fail, because the 3 captured spans don't belong to 1 trace, as they should.

If you comment out the `@Asynchronous` annotation on the `HelloClient` interface and replace the `when().get("/client").then().statusCode(...).body(...)` with just `when().get("/client").then()` in the `ReproducerTest`, you will see that there's just 1 trace that contains all 3 spans. Of course, the fallback behavior doesn't work (that's why we had to remove the `statusCode()` and `body()` checks).

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:
```
lthon@argondie:~$ uname -a
Linux argondie 4.15.0-96-generic #97-Ubuntu SMP Wed Apr 1 03:25:46 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
 - Output of `java -version`: 
```
lthon@argondie:~$ java -version
openjdk version "1.8.0_242"
OpenJDK Runtime Environment (build 1.8.0_242-8u242-b08-0ubuntu3~18.04-b08)
OpenJDK 64-Bit Server VM (build 25.242-b08, mixed mode)
```
 - GraalVM version (if different from Java): N/A
 - Quarkus version or git rev: `master`, specifically `93c5d658a135d17c320100aa7fae9debcb44ca93`
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
```
lthon@argondie:~$ mvn -version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /home/lthon/software/apache-maven
Java version: 1.8.0_242, vendor: Private Build, runtime: /usr/lib/jvm/java-8-openjdk-amd64/jre
Default locale: cs_CZ, platform encoding: UTF-8
OS name: "linux", version: "4.15.0-96-generic", arch: "amd64", family: "unix"
```

**Comment by Ladicek** [comment]
I'm not sure exactly who's responsibility this is, and I can see a long pingpong between RestClient and Fault Tolerance happening, but for start, I'm CC'ing:

- @asoldano for RestClient (because it's implemented in RESTEasy)
- @pavolloffay for OpenTracing (but that is just a victim here, not a suspect)
- @FroMage as a context propagation expect (as conprop is what doesn't entirely work here; it's not MP Context Propagation, but the MP RestClient's own API for conprop)

I could also CC myself for Fault Tolerance, but that's kinda unnecessary :-)

**Comment by Ladicek** [comment]
I have just found a ... well, not sure if solution, but at least a workaround: use MicroProfile Context Propagation (because SmallRye Fault Tolerance understands that) to make sure that the `Collection<AsyncInvocationInterceptor>` survives the move to a different thread. It's a relatively simple patch to RESTEasy: https://github.com/Ladicek/Resteasy/commit/a5d56b8669

It adds an `<optional>true</optional>` dependency on `microprofile-context-propagation-api` to `resteasy-client-microprofile`, which should be fine -- it isn't propagated transitively, it's there only for compilation. When MP ConProp isn't present at runtime, the 1 added class is just dead weight, but when MP ConProp is present, that 1 class makes sure that the thread-local `Collection<AsyncInvocationInterceptor>` moves to a different thread together with execution. I wonder if that would be acceptable to RESTEasy folks.

**Comment by asoldano** [comment]
Thanks @Ladicek , this looks interesting, I think you should open up a PR against RESTEasy with your changes for additional review. /cc @ronsigal.

**Comment by FroMage** [comment]
Hi. Sorry but there are so many things wrong with how MP does context propagation here that it has to be put to a halt. I've seen this issue for months and complained and it's still here and MP REST client and MP FT continued with the faulty approach. We should fix it once and for all.

> it also has the @Asynchronous annotation (that's per MP Fault Tolerance; if it didn't have @Asynchronous, MP Fault Tolerance would treat the method call as synchronous and fault tolerance behaviors would only work for the method invocation itself and not for all the async actions that happen after the method returns).

This is wrong. The semantics of JAX-RS server endpoints when it returns a `CS` (or in RESTEasy any reactive type) is that the main method is invoked on the normal thread and the request is suspended (effectively async).

MP REST Client understands this, and so should MP-FT. The `@Asynchronous` annotation forces execution to a worker thread that should only be justifiable for non-reactive code. MP-FT should understand MP-REST definitions of asynchronousy based on the reactive return type and that annotation should not be needed for it to function, because it has side-effects incompatible with the semantics of MP-REST/JAX-RS.

> MP Fault Tolerance moves the execution to another thread, and then MP RestClient moves the execution to yet another thread

This is not strictly true in the sense that MP REST Client only moves execution of the result of the HTTP call to another thread. Thus it does not require a thread pool and can function on the IO thread, unlike MP-FT.

> I was able to trace this down to the RESTEasy implementation of MP RestClient, which collects AsyncInvocationInterceptors (MP RestClient way of doing context propagation) in a ThreadLocal variable (see org.jboss.resteasy.microprofile.client.async.AsyncInvocationInterceptorHandler). The context is captured properly, but it not activated, because the activation code runs on a different thread and won't find the Collection<AsyncInvocationInterceptor>, because it was stored into a ThreadLocal on a different thread.

Yes. This is wrong. This is an attempt of MP REST Client at not depending on MP-CP and paying the price for it. This thread-local is set and used in non-atomic ways which pretty much ensures that things will go wrong.

This API should be removed from MP, in favour of using MP-CP. That's the real fix. Unfortunately, it appears to not be done due to politics, because MP-CP is not in the base platform. I should raise this to @kenfinnigan's attention that we're sacrificing correctness here by duplicating features across the specs for no good reason. We should fix this.

We can work around it by fixing how that thread-local is used in resteasy-mp-client, for sure, as this is the fastest way to get a fix (unless we can fix SR-FT to not spawn a thread for async-return-type methods). But we should also work on fixing the underlying issues that I've explained above.

**Comment by FroMage** [comment]
> I have just found a ... well, not sure if solution, but at least a workaround: use MicroProfile Context Propagation (because SmallRye Fault Tolerance understands that) to make sure that the Collection<AsyncInvocationInterceptor> survives the move to a different thread. It's a relatively simple patch to RESTEasy: Ladicek/Resteasy@a5d56b8

This would only fix it if the implementation of resteasy-mp-client uses MP-CP to propagate contexts in its client, no? I don't believe this to be the case ATM. MP-CP is automatic if using Mutiny or RxJava, but not if using `CompletionStage` so unless there's manual code in resteasy-mp-client to use MP-CP to wrap the `CompletionStage` it uses for the client, this won't work.

It also is the wrong way to fix things in RESTEasy, I believe, because RESTEasy so far supports MP-CP but doesn't require it. The RESTEasy way of doing things is that thread-locals such as what `ResteasyContext` supports are captured in the internals everywhere a request can be suspended, and internally restored every time the request can be resumed, so as to hide this from users and not depend on MP-CP (for RESTEasy internals).

MP-CP will bring you other contexts, such as transactions and CDI, and that's not required for RESTEasy internals to function.

I believe the most idiomatic way to fix this in resteasy-mp-client would be to properly capture that thread-local and tie it to the request.

Now, to be frank, I don't even know why we need a thread-local for it. It doesn't seem right at first glance.

I'd try a fix myself, but I've never been able to really understand what this piece of code does, because it's too weird from the PoV of CP. And I lack a decent test to verify it.

**Comment by FroMage** [comment]
OK I've just re-read the code and the entire weirdness of this thread-local is due to the fact that the resteasy-mp-client  uses a decorated `ExecutorService` (huh, MP-CP has one too) per client builder (for all requests), but the context is supposed to be specific per request, so there's an impedance mismatch where we're trying to pass a list of context restorers (`AsyncInvocationInterceptor`) from a specific request call to a global `ExecutorService` in the hopes that its next call will guess that it was really doing this current request.

This is at best unsafe and won't likely work if multiple threads access the same client at the same time because it's very likely calls to the `ExecutorService`'s thread-local will cross.

We either need an `ExecutorService` wrapper per request (not per client), or not rely on it for context propagation.

**Comment by Ladicek** [comment]
> The `@Asynchronous` annotation forces execution to a worker thread that should only be justifiable for non-reactive code.

I have tried several times to convince the MP FT folks that `@Asynchronous` doesn't really work on the event loop. So far, I haven't been successful. I eventually decided that I first need to prototype a solution (https://github.com/smallrye/smallrye-fault-tolerance/issues/192), otherwise I don't have a chance.

Unfortunately, the MP FT spec currently mandates that the method is called on an extra thread. So that's what we do.

(Yes, we can fix it. No, I didn't have the time yet :-) )

> This would only fix it if the implementation of resteasy-mp-client uses MP-CP to propagate contexts in its client, no?

With the "fix" I proposed, the MP RestClient implementation doesn't do anything, because indeed it doesn't use MP ConProp. But SmallRye Fault Tolerance does. SRye FT uses a `ManagedExecutor` for running all the `@Asynchronous` actions, and so the `AsyncInvocationInterceptorThreadContextProvider` is used to propagate the `Collection<AsyncInvocationInterceptor>` from the `ThreadLocal` in original thread to the `ThreadLocal` in the new (required by MP FT) thread.

**Comment by FroMage** [comment]
> I have tried several times to convince the MP FT folks that @Asynchronous doesn't really work on the event loop. So far, I haven't been successful. I eventually decided that I first need to prototype a solution (smallrye/smallrye-fault-tolerance#192), otherwise I don't have a chance.

It appears to me that you're trying to make it so `@Asyncrhonous` doesn't spawn a new thread in some cases. I rather think that `@Asynchronous` should not be needed at all when returning an async type.

> With the "fix" I proposed, the MP RestClient implementation doesn't do anything, because indeed it doesn't use MP ConProp. But SmallRye Fault Tolerance does. SRye FT uses a ManagedExecutor for running all the @Asynchronous actions, and so the AsyncInvocationInterceptorThreadContextProvider is used to propagate the Collection<AsyncInvocationInterceptor> from the ThreadLocal in original thread to the ThreadLocal in the new (required by MP FT) thread.

OK, now I get it.

**Comment by Ladicek** [comment]
> It appears to me that you're trying to make it so `@Asyncrhonous` doesn't spawn a new thread in some cases. I rather think that `@Asynchronous` should not be needed at all when returning an async type.

That makes sense. (For the record, for the prototype I was thinking about, I wanted to create a new annotation `@VertxAsynchronous`, so that it's easy to verify and doesn't break existing behavior.)

Actually, I think it should be rather easy to implement async FT for `CompletionStage`-returning methods without moving them to a new thread. I'll try.

**Comment by Ladicek** [comment]
OK, so I added support for async FT for `CompletionStage`-returning methods to SmallRye Fault Tolerance, it was relatively straightforward and it works (https://github.com/smallrye/smallrye-fault-tolerance/pull/218). I'm not sure if we can enable it unconditionally, because we have this nasty bug: https://github.com/smallrye/smallrye-fault-tolerance/issues/210 But it's also a possible fix for the present problem.

**Comment by FroMage** [comment]
I've added a draft PR for my fix proposal at https://github.com/resteasy/Resteasy/pull/2359

This gets us rid of the thread-local and should be sane. But I still think context propagation should not be done this way.

**Comment by ronsigal** [comment]
So, we have ... three solutions?

1. https://github.com/Ladicek/Resteasy/commit/a5d56b8669
2. https://github.com/resteasy/Resteasy/pull/2359
3. https://github.com/smallrye/smallrye-fault-tolerance/pull/218

Note that @kabir has been working on getting context propagation to work in Wildfly, including in RESTEasy, which would involve porting the resteasy-context-progagation module from RESTEasy 4.x to 3.x. It has a ThreadContextProvider that supports just the core RESTEasy stuff. but wouldn't adding @Ladicek's AsyncInvocationInterceptorThreadContextProvider be simple and symmetric?

Btw, I think he's still in Extension Pack hell, so RESTEasy context propagation may or may not come to WF 20.

**Comment by Ladicek** [comment]
1. https://github.com/Ladicek/Resteasy/commit/a5d56b8669 is a worse idea than https://github.com/resteasy/Resteasy/pull/2359 and I fully support getting https://github.com/resteasy/Resteasy/pull/2359 merged.
2. https://github.com/smallrye/smallrye-fault-tolerance/pull/218 is sort of orthogonal, and if I make the RestClient method `@Asynchronous` (which is arguably a bad idea, but current MP Fault Tolerance spec requires it), it still requires the previous patch.

**Comment by rsvoboda** [comment]
@ronsigal / @FroMage any updates or plans regarding this issue?

**Comment by FroMage** [comment]
Well https://github.com/resteasy/Resteasy/pull/2359 got merged. Not sure what else @Ladicek needs?

**Comment by rsvoboda** [comment]
Looks like this was fixed in Quarkus 1.4.0.Final, I see the issue with 1.3.4.Final.
